src/main/java/org/elasticsearch/search/aggregations/reducers/movavg/models/MovAvgModel.java

```diff
@@ -0,0 +1,100 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.reducers.movavg.models;
+
+import com.google.common.collect.EvictingQueue;
+import org.elasticsearch.ElasticsearchIllegalArgumentException;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.*;
+
+public abstract class MovAvgModel {
+
+    /**
+     * Returns the next value in the series, according to the underlying smoothing model
+     *
+     * @param values    Collection of numerics to movingAvg, usually windowed
+     * @param <T>       Type of numeric
+     * @return          Returns a double, since most smoothing methods operate on floating points
+     */
+    public abstract <T extends Number> double next(Collection<T> values);
+
+    /**
+     * Predicts the next `n` values in the series, using the smoothing model to generate new values.
+     * Default prediction mode is to simply continuing calling <code>next()</code> and adding the
+     * predicted value back into the windowed buffer.
+     *
+     * @param values            Collection of numerics to movingAvg, usually windowed
+     * @param numPredictions    Number of newly generated predictions to return
+     * @param <T>               Type of numeric
+     * @return                  Returns an array of doubles, since most smoothing methods operate on floating points
+     */
+    public <T extends Number> double[] predict(Collection<T> values, int numPredictions) {
+        double[] predictions = new double[numPredictions];
+
+        // If there are no values, we can't do anything.  Return an array of NaNs.
+        if (values.size() == 0) {
+            return emptyPredictions(numPredictions);
+        }
+
+        // special case for one prediction, avoids allocation
+        if (numPredictions < 1) {
+            throw new ElasticsearchIllegalArgumentException("numPredictions may not be less than 1.");
+        } else if (numPredictions == 1){
+            predictions[0] = next(values);
+            return predictions;
+        }
+
+        // nocommit
+        // I don't like that it creates a new queue here
+        // The alternative to this is to just use `values` directly, but that would "consume" values
+        // and potentially change state elsewhere.  Maybe ok?
```

Comment:
I'd much rather like be a bit wasteful than have the risk to introduce side-effects elsewhere, as it could be super hard to diagnose/debug.